python/plugins/processing/algs/qgis/EliminateSelection.py

```diff
@@ -0,0 +1,227 @@
+
+# This will get replaced with a git SHA1 when you do a git archive
+
+__revision__ = '$Format:%H$'
+
+import os
+
+from qgis.PyQt.QtGui import QIcon
+
+from qgis.core import QgsFeatureRequest, QgsFeature, QgsGeometry
+
+from processing.core.GeoAlgorithm import GeoAlgorithm
+from processing.core.GeoAlgorithmExecutionException import GeoAlgorithmExecutionException
+from processing.core.ProcessingLog import ProcessingLog
+from processing.core.parameters import ParameterVector
+from processing.core.parameters import ParameterSelection
+from processing.core.outputs import OutputVector
+from processing.tools import dataobjects, vector
+
+pluginPath = os.path.split(os.path.split(os.path.dirname(__file__))[0])[0]
+
+
+class EliminateSelection(GeoAlgorithm):
+
+    INPUT = 'INPUT'
+    OUTPUT = 'OUTPUT'
+    MODE = 'MODE'
+
+    MODE_LARGEST_AREA = 0
+    MODE_SMALLEST_AREA = 1
+    MODE_BOUNDARY = 2
+
+    def __init__(self):
+        GeoAlgorithm.__init__(self)
+        self.showInToolbox = True
+        self.showInModeler = True
+
+    def getIcon(self):
+        return QIcon(os.path.join(pluginPath, 'images', 'ftools', 'eliminate.png'))
+
+    def defineCharacteristics(self):
+        self.name, self.i18n_name = self.trAlgorithm('Eliminate selected polygons')
+        self.group, self.i18n_group = self.trAlgorithm('Vector geometry tools')
+
+        self.modes = [self.tr('Largest area'),
+                      self.tr('Smallest Area'),
+                      self.tr('Largest common boundary')]
+
+        self.addParameter(ParameterVector(self.INPUT,
+                                          self.tr('Input layer'), [dataobjects.TYPE_VECTOR_POLYGON]))
+        self.addParameter(ParameterSelection(self.MODE,
+                                             self.tr('Merge selection with the neighbouring polygon with the'),
+                                             self.modes))
+        self.addOutput(OutputVector(self.OUTPUT, self.tr('Eliminated'), datatype=[dataobjects.TYPE_VECTOR_POLYGON]))
+
+    def processAlgorithm(self, feedback):
+        inLayer = dataobjects.getObjectFromUri(self.getParameterValue(self.INPUT))
+        boundary = self.getParameterValue(self.MODE) == self.MODE_BOUNDARY
+        smallestArea = self.getParameterValue(self.MODE) == self.MODE_SMALLEST_AREA
+        processLayer = vector.duplicateInMemory(inLayer)
+
+        if inLayer.selectedFeatureCount() == 0:
+            ProcessingLog.addToLog(ProcessingLog.LOG_WARNING,
+                                   self.tr('%s: (No selection in input layer "%s")' % (self.commandLineName(), self.getParameterValue(self.INPUT))))
+
+        # Keep references to the features to eliminate
+        featToEliminate = []
+
+        for aFeat in inLayer.selectedFeatures():
+            featToEliminate.append(aFeat)
+
+        # select and delete the features in ProcessLayer that are selected in inLayer
+        idsToEliminate = []
+
+        for aFeat in featToEliminate:
```

Comment:
I wouldn't invert the selection - that requires a full iteration through the layer to perform the invert, then a second iteration through the newly selected features to save them to the new layer. You're better off looping through all the features in the layer and adding them to the new layer, and just skipping ones which were in the original selection